internal/proxy/engines/objectproxycache.go

```diff
@@ -172,6 +179,214 @@ func FetchViaObjectProxyCache(r *model.Request, client model.Client, apc *config
 
 }
 
+// FetchAndRespondViaObjectProxyCache Fetches an object from Cache or Origin (on miss), writes the object to the cache, and returns the object to the caller
+func FetchAndRespondViaObjectProxyCache(r *model.Request, w http.ResponseWriter, client model.Client, noLock bool) {
+
+	oc := context.OriginConfig(r.ClientRequest.Context())
+	pc := context.PathConfig(r.ClientRequest.Context())
+	cache := context.CacheClient(r.ClientRequest.Context())
+
+	key := oc.Host + "." + DeriveCacheKey(r, pc, "")
+
+	cpReq := GetRequestCachingPolicy(r.Headers)
+	pcfResult, pcfExists := Reqs.Load(key)
+	if pcfExists || cpReq.NoCache {
+		// if the client provided Cache-Control: no-cache or Pragma: no-cache header, the request is proxy only.
+		start := time.Now()
+		// This will use the header from the original response
+		resp := ProxyRequest(r, w)
+		recordOPCResult(r, tc.LookupStatusProxyOnly, resp.StatusCode, r.URL.Path, time.Since(start).Seconds(), resp.Header)
+
+		locks.Acquire(key)
+		cache.Remove(key)
+		locks.Release(key)
+		return
+	}
+
+	hasINMV := cpReq.IfNoneMatchValue != ""
```

Comment:
not a fan of repeating a lot of the caching policy checks from the basic Fetch, but i understand it's a lot to try and unwind. Good enough for now, but let's make sure to get this more DRY when there are cycles to revisit it.